src/hotspot/share/asm/register.hpp

```diff
@@ -60,263 +61,18 @@ const type name = ((type)value)
-                                                                                 && i != j && i !=k
-                                                                                           && j !=k,
-    "registers must be different: a=" INTPTR_FORMAT ", b=" INTPTR_FORMAT
-    ", c=" INTPTR_FORMAT ", d=" INTPTR_FORMAT ", e=" INTPTR_FORMAT
-    ", f=" INTPTR_FORMAT ", g=" INTPTR_FORMAT ", h=" INTPTR_FORMAT
-    ", i=" INTPTR_FORMAT ", j=" INTPTR_FORMAT ", k=" INTPTR_FORMAT "",
-    p2i(a), p2i(b), p2i(c), p2i(d), p2i(e), p2i(f), p2i(g), p2i(h), p2i(i), p2i(j), p2i(k)
-  );
-}
-
-inline void assert_different_registers(
-  AbstractRegister a,
-  AbstractRegister b,
-  AbstractRegister c,
-  AbstractRegister d,
-  AbstractRegister e,
-  AbstractRegister f,
-  AbstractRegister g,
-  AbstractRegister h,
-  AbstractRegister i,
-  AbstractRegister j,
-  AbstractRegister k,
-  AbstractRegister l
-) {
-  assert(
-    a != b && a != c && a != d && a != e && a != f && a != g && a != h && a != i && a != j && a !=k && a !=l
-           && b != c && b != d && b != e && b != f && b != g && b != h && b != i && b != j && b !=k && b !=l
-                     && c != d && c != e && c != f && c != g && c != h && c != i && c != j && c !=k && c !=l
-                               && d != e && d != f && d != g && d != h && d != i && d != j && d !=k && d !=l
-                                         && e != f && e != g && e != h && e != i && e != j && e !=k && e !=l
-                                                   && f != g && f != h && f != i && f != j && f !=k && f !=l
-                                                             && g != h && g != i && g != j && g !=k && g !=l
-                                                                       && h != i && h != j && h !=k && h !=l
-                                                                                 && i != j && i !=k && i !=l
-                                                                                           && j !=k && j !=l
-                                                                                                    && k !=l,
-    "registers must be different: a=" INTPTR_FORMAT ", b=" INTPTR_FORMAT
-    ", c=" INTPTR_FORMAT ", d=" INTPTR_FORMAT ", e=" INTPTR_FORMAT
-    ", f=" INTPTR_FORMAT ", g=" INTPTR_FORMAT ", h=" INTPTR_FORMAT
-    ", i=" INTPTR_FORMAT ", j=" INTPTR_FORMAT ", k=" INTPTR_FORMAT
-    ", l=" INTPTR_FORMAT "",
-    p2i(a), p2i(b), p2i(c), p2i(d), p2i(e), p2i(f), p2i(g), p2i(h), p2i(i), p2i(j), p2i(k), p2i(l)
-  );
+template<typename R, typename... Rx>
+inline void assert_different_registers(R first_register, Rx... more_registers) {
+#ifdef ASSERT
+  R regs[] = { first_register, more_registers... };
+  // Sort regs, so any equal entries are adjacent.
+  struct Compare { ptrdiff_t operator()(R x, R y) const { return x - y; } };
+  QuickSort::sort(regs, ARRAY_SIZE(regs), Compare(), false);
```

Comment:
Sorting is kind of overkill.  Maybe I should have just use the obvious nested loop?  What do reviewers think?

Reply:
Answering my own question, sorting imposes additional requirements on the type that might require adjustment for JDK-8270140.  So I'm inclined to switch to the nested loop that only requires equality testing as before.

Reply:
Well, a nested loop is order n^2 but n is always going to be pretty small (the old code allowed for 12 but cases with n > 6 are rare). So I see no great perf problem with using a nested loop.

Reply:
... bearing in mind that quicksort is n.log(n) at best.

Reply:
> Sorting is kind of overkill. Maybe I should have just use the obvious nested loop? What do reviewers think?

No strong preference on my side. I'm fine with it either way.

Reply:
Updated to use nested loops.